### ISSUE TITLE
rustix keeps rebuilding in my dependency tree.

### ISSUE THREAD

**Issue by rukai** [bytecodealliance/rustix#575]
My application has rustix in its dependency tree due to clap:
```
├── clap v4.1.11
│   ├── bitflags v2.0.2
│   ├── clap_derive v4.1.9 (proc-macro)
│   │   ├── heck v0.4.1
│   │   ├── proc-macro-error v1.0.4
│   │   │   ├── proc-macro-error-attr v1.0.4 (proc-macro)
│   │   │   │   ├── proc-macro2 v1.0.52 (*)
│   │   │   │   └── quote v1.0.26 (*)
│   │   │   │   [build-dependencies]
│   │   │   │   └── version_check v0.9.4
│   │   │   ├── proc-macro2 v1.0.52 (*)
│   │   │   ├── quote v1.0.26 (*)
│   │   │   └── syn v1.0.109 (*)
│   │   │   [build-dependencies]
│   │   │   └── version_check v0.9.4
│   │   ├── proc-macro2 v1.0.52 (*)
│   │   ├── quote v1.0.26 (*)
│   │   └── syn v1.0.109 (*)
│   ├── clap_lex v0.3.3
│   │   └── os_str_bytes v6.5.0
│   ├── is-terminal v0.4.5
│   │   ├── io-lifetimes v1.0.9
│   │   │   └── libc v0.2.140
│   │   └── rustix v0.36.11
│   │       ├── bitflags v1.3.2
│   │       ├── io-lifetimes v1.0.9 (*)
│   │       ├── libc v0.2.140
│   │       └── linux-raw-sys v0.1.4
│   ├── once_cell v1.17.1
│   ├── strsim v0.10.0
│   └── termcolor v1.2.0
```

I have no idea why but sometimes rustix will just randomly trigger a rebuild.
```
   Compiling rustix v0.36.11
    Checking is-terminal v0.4.5
    Checking clap v4.1.11
    Checking inferno v0.11.15
    Checking shotover-proxy v0.1.9 (/home/rukai2/Projects/Crates/shotover/shotover-proxy/shotover-proxy)
    Checking test-helpers v0.1.0 (/home/rukai2/Projects/Crates/shotover/shotover-proxy/test-helpers)
    Finished dev [unoptimized + debuginfo] target(s) in 6.51s
```

I've never seen this for any other dependency before.

If I observe anything else that might explain whats going on I'll add it to this issue.

**Comment by sunfishcode** [comment]
Are you using an IDE, such as IntelliJ or another? Are you using `cargo watch` or similar tool which run commands automatically?

**Comment by rukai** [comment]
I have vs code open + rust analyzer open and [bacon](https://github.com/Canop/bacon) open at the same time.

**Comment by rukai** [comment]
Looking through the rust-analyzer vs code config I saw:
![image](https://user-images.githubusercontent.com/5120858/228404775-256563e7-8b8a-4525-9231-5f4bda824b3e.png)

Which corresponds to https://github.com/bytecodealliance/rustix/blob/1a9d8dacb06c62c028c0936031f3122a649722d2/build.rs#L153

But also in that case wouldnt https://github.com/bytecodealliance/rustix/blob/1a9d8dacb06c62c028c0936031f3122a649722d2/build.rs#L154 cause a rebuild every time I switch between release and debug? Which I've observed is not the case.

Cargo must have its own internal logic for skipping rebuilds before deferring to the build.rs `rerun-if` logic.

**Comment by rukai** [comment]
https://doc.rust-lang.org/cargo/reference/build-scripts.html#cargorerun-if-env-changedname

> Note that the environment variables here are intended for global environment variables like CC and such, it is not necessary to use this for environment variables like TARGET that Cargo sets.

Oh this looks useful: https://doc.rust-lang.org/cargo/faq.html#why-is-cargo-rebuilding-my-code

**Comment by rukai** [comment]
I discovered that cargo recently had reasons for rebuilds added to its output when `--verbose` is used.

Through doing this I was able to catch RUSTC_WRAPPER as the cause:
![image](https://user-images.githubusercontent.com/5120858/228408716-22299459-86f9-411e-8244-5f074d7a300f.png)

**Comment by VorpalBlade** [comment]
I'm not sure if this is the same issue, as I get a slightly different info with `cargo build --verbose`:
```
Dirty rustix v0.37.4: the dependency build_script_build was rebuilt (1680288365.394171857s, 7m 15s after last build at 1680287930.340354402s)
```
Unfortunately that doesn't tell which env var it is. I too am using VSCode with rust-analyzer.

**Comment by theHausdorffMetric** [comment]
> I'm not sure if this is the same issue, as I get a slightly different info with `cargo build --verbose`:
> 
> ```
> Dirty rustix v0.37.4: the dependency build_script_build was rebuilt (1680288365.394171857s, 7m 15s after last build at 1680287930.340354402s)
> ```
> 
> Unfortunately that doesn't tell which env var it is. I too am using VSCode with rust-analyzer.

Same issue here (rustix keeps rebuilding) with samecargo verbose output, but using helix editor with rust-analyzer.

**Comment by rudyhb** [comment]
Same issue for me with CLion + rust plugin

**Comment by sunfishcode** [comment]
Could people still seeing this issue please include what version of rustix they're seeing it in?

I am currently unable to reproduce this, so any extra information you might be able to provide about the circumstances you see this in could be helpful. Thanks!

**Comment by rukai** [comment]
I havent seen it in a while, I'll report back if I do see it again.

**Comment by rudyhb** [comment]
sorry about that, issue seems to be fixed after updating from v0.36.11 to v0.37.11

**Comment by theHausdorffMetric** [comment]
> sorry about that, issue seems to be fixed after updating from v0.36.11 to v0.37.11

Many thx - preliminary testing v0.37.11 in my case (helix editor with rust-analyzer) no longer shows the problem :-)

**Comment by rukai** [comment]
> I havent seen it in a while, I'll report back if I do see it again.

I havent seen the problem since then so I'm going to go ahead and close the issue.
Thanks for resolving it!
